### ISSUE TITLE
Diff PDFs by converting them to HTML and diffing that

### ISSUE THREAD

**Issue by Mr0grog** [edgi-govdata-archiving/web-monitoring-diff#9]
This should not necessarily *replace* [a differ for actual PDF content](https://github.com/edgi-govdata-archiving/web-monitoring/issues/36), but it could potentially be a lot more useful when it works well: instead of trying to diff two PDF files, convert the PDF to HTML (there are at least a few open-source libraries for this) and feed that through the HTML differ.

Not sure what the right name for this is.

Lizz brought this up in Slack and, though I remember having a short discussion about the idea before, I can’t find anywhere we’ve written it down, hence this issue.

**Comment by AdeelTahir-SE** [comment]
Yeah @Mr0grog , I think converting PDFs to HTML and then using the existing HTML differ could significantly reduce complexity and allow us to reuse the current pipeline.

However, this approach may introduce issues such as loss of layout, formatting inconsistencies, and potential omission of meaningful changes especially in complex or non-text-based PDFs.

An alternative approach would be to extract structured text directly from PDFs and compute diffs on that. While this could provide more accurate results, it would likely require more implementation effort compared to the HTML-based approach.

I’m currently exploring these approaches and evaluating their trade-offs (accuracy vs complexity). I’ll share more concrete findings once I have some results.

**Comment by AdeelTahir-SE** [comment]
i saw article above mentioned by @Mr0grog and also from the research i did i found out that threre is no library or tool which can extract all the information from pdfs
so what i beleive we can do is create a pipeline (Multi-modal PDF information extraction pipeline):
=> use different libraries and packages instead of any one library each specializing in extracting specific information like (text, image and content in images) 
=>extract information through each and aggregate information from each into a json format and perfomr diff on that unified representation
This might help improve coverage and accuracy compared to a single-method approach, though it does introduce additional complexity. Now the final information will be used in diffing.
Also you can see in the images showing proposed architecture

1. **_Information Extraction Pipeline Image:_**

<img width="501" height="372" alt="Image" src="https://github.com/user-attachments/assets/7c46266d-91a9-49f2-ac02-1c14045be59c" />

2. **_Complete Architecture:_**
 
<img width="912" height="260" alt="Image" src="https://github.com/user-attachments/assets/e8287939-8c1f-49f9-9a80-08868778d8dc" />

**Comment by AdeelTahir-SE** [comment]
I am working on the prototype will share the details soon

**Comment by Mr0grog** [comment]
@AdeelTahir-SE great, I am curious to see what you come up with! This is an interesting idea, and I’m excited someone is taking up the PDF work again.

In general, I’d say PDF is a really messy and inconsistent format when it comes to trying to extract text or data in any meaningful form—no matter what you do here, it’s going to be lossy and imperfect.

I think my gut reaction is that, because of the above issues, this approach is likely to get quite complicated (still happy to see the idea explored in case it does work!). If you find your experiments aren’t working too well, you might try something more like the approaches discussed in https://github.com/edgi-govdata-archiving/web-monitoring/issues/36, where existing libraries for highlighting or drawing a box around changed areas of the PDF are explored.

Separately, even a really dumb and basic comparison, for example comparing the number of pages and document title or other metadata, would be really useful here!
